lib/secrets/secrets.go

```diff
@@ -229,3 +279,155 @@ func PrepareSSHPrivateKey(keyContent string) (string, error) {
+	stream.XORKeyStream(plain, ciphertext)
+	d, err := decodeBase64(string(plain))
+	if err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+// function used to write the secret values to given file in toml format
+// stores the name of the algorithm , iv and the ciphertext
+func writeToml(algorithm string, iv string, ciphertext string, path string) error {
+	secretData := SecretKeyToml{
+		Algorithm:  algorithm,
+		IV:         iv,
+		Ciphertext: ciphertext,
+	}
+
+	buf := new(bytes.Buffer)
+	if err := toml.NewEncoder(buf).Encode(secretData); err != nil {
+		log.Fatal(err)
+	}
+
+	err := fileutils.AtomicWrite(path, bytes.NewReader(buf.Bytes()), fileutils.WithAtomicWriteFileMode(0700))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// takes in the secret value and encrypts it
+// return encrypted cipher, iv and error
+func getEncryptedData(data []byte, basePath string) ([]byte, []byte, error) {
+	// we create the encryption key if it doesnot exist
+	var secretKey []byte
+	keyExists, err := keyExists(basePath)
+	if err != nil {
+		return nil, nil, err
+	}
+	if !keyExists {
+		secretKey, err = createEncryptionKey(basePath)
+		if err != nil {
+			return nil, nil, err
+		}
+	} else {
+		path := filepath.Join(basePath, "key")
+		secretKey, err = ioutil.ReadFile(path)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "could not read secrets key (unexpected error)")
+		}
+	}
+	encryptedData, iv, err := encrypt(secretKey, data)
+	if err != nil {
+		return nil, nil, err
+	}
+	return encryptedData, iv, nil
+}
+
+// gets the secret from the toml file and decrypts it
+// returns the decrypted data or error
+func getDecryptedData(basePath string, ret []byte) ([]byte, error) {
+	keyPath := filepath.Join(basePath, "key")
+	key, err := ioutil.ReadFile(keyPath)
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+	var secretData SecretKeyToml
+	_, err = toml.Decode(string(ret), &secretData)
+	if err != nil {
+		return nil, err
+	}
+
+	// coverts ciphertext to byte from base64
+	dcdCipher, err := decodeBase64(secretData.Ciphertext)
+	if err != nil {
+		return nil, err
+	}
+	// coverts iv to byte from base64
+	dcdIv, err := decodeBase64(secretData.IV)
+	if err != nil {
+		return nil, err
+	}
+
+	decrypted, err := decrypt(block, dcdCipher, dcdIv)
+	if err != nil {
+		return nil, err
+	}
+	return decrypted, nil
+}
+
+// Creates the Key used for the secret values' encryption
+// we generate a random byte array of 32 bytes required for the AES256 encryption
+// returns the key
+func createEncryptionKey(basePath string) ([]byte, error) {
+	secretKey, err := GenerateRandomBytes(32)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not generate random secret for encryption")
+	}
+
+	// creating path to store the secret key
+	secretKeyPath := filepath.Join(basePath, "key")
+	s := bytes.NewReader(secretKey)
+	// writing the key to a file
+	err = fileutils.AtomicWrite(secretKeyPath, s, fileutils.WithAtomicWriteFileMode(0700))
```

Comment:
I think you need to set the owner of the file

Reply:
setting owner now